## 기초데이터과학 (01분반)

## Lab 03. 데이터 전처리 실습

### 진행 방식

- **문제 단위**로 진행하며, 문제마다 제한시간이 있음
- 코드 틀에서 `None` 으로 비어 있는 부분만 채우면 됨
- `print` 문은 이미 작성되어 있으므로 수정 불필요
- 실행 결과가 아래 **기대 출력**과 같으면 정답
- 제한시간이 끝나면 다 같이 풀이 진행. 못 풀었으면 그때 고치면 됨

### 실습 내용

- `loc[ ]` 과 `iloc[ ]` 의 차이
- 조건으로 행 추출하기 (`query( )`, 대괄호 방식)
- 정렬과 인덱스 정리
- 실제 데이터에 전처리 적용하기

---

## 문제 1. `loc[ ]` 과 `iloc[ ]` 구분하기 (제한시간 10분)

- 아래 셀을 먼저 실행할 것
- 어느 매장에서 어떤 상품이 몇 개 팔렸는지 기록한 데이터

In [1]:
import pandas as pd

sales = pd.DataFrame({
    'store':    ['서울', '서울', '부산', '부산', '대구', '대구', '광주', '광주'],
    'item':     ['노트북', '마우스', '노트북', '키보드', '마우스', '모니터', '노트북', '모니터'],
    'price':    [1200, 25, 1150, 80, 30, 300, 1300, 280],
    'quantity': [3, 40, 5, 12, 35, 8, 2, 10]
})
print(sales)

  store item  price  quantity
0    서울  노트북   1200         3
1    서울  마우스     25        40
2    부산  노트북   1150         5
3    부산  키보드     80        12
4    대구  마우스     30        35
5    대구  모니터    300         8
6    광주  노트북   1300         2
7    광주  모니터    280        10


### 아래 셀을 실행하기 전에 생각해 볼 것

- 다음 두 줄은 **똑같은 숫자** 0과 3을 사용함

```python
sales.iloc[0:3]
sales.loc[0:3]
```

- 두 결과의 **행 개수**는 같을까, 다를까?
- 다르다면 각각 몇 행이 나오며, 그 이유는 무엇일까?

> 나의 예측:


In [ ]:
print("iloc[0:3]")
print(sales.iloc[0:3])
print()
print("loc[0:3]")
print(sales.loc[0:3])

#### 1-1. `iloc[ ]` 을 이용하여 **처음 두 행의 `store`, `item` 열**만 추출하시오.
#### 1-2. `loc[ ]` 을 이용하여 **`price` 가 1000 이상인 행의 `item`, `price` 열**만 추출하시오.
#### 1-3. `quantity` 기준으로 내림차순 정렬한 결과를 `sales_sorted` 에 저장하시오.
#### 1-4. `sales_sorted` 에서 `loc[0]` 과 `iloc[0]` 이 각각 어떤 행을 가리키는지 출력하시오.

> 힌트 1-1: `iloc[ ]` 은 행도 열도 **위치 번호**로 지정
> 
> 힌트 1-2: `loc[행 조건, 열 목록]` 형태로 쉼표를 이용하여 한 번에 작성
>
> 힌트 1-4: 정렬을 해도 인덱스(행의 이름)는 행을 따라다님

In [ ]:
# 1-1. 처음 두 행의 store, item 열
a = None               # TODO

# 1-2. price 가 1000 이상인 행의 item, price 열
b = None               # TODO

# 1-3. quantity 내림차순 정렬
sales_sorted = None    # TODO

# 1-4. 정렬 결과에서 loc[0] 과 iloc[0]
c = None               # TODO
d = None               # TODO

print("[1-1]")
print(a)
print()
print("[1-2]")
print(b)
print()
print("[1-3]")
print(sales_sorted)
print()
print("[1-4] loc[0] ->", c['store'], c['item'])
print("[1-4] iloc[0] ->", d['store'], d['item'])

**기대 출력**

```
[1-1]
  store item
0    서울  노트북
1    서울  마우스

[1-2]
  item  price
0  노트북   1200
2  노트북   1150
6  노트북   1300

[1-3]
  store item  price  quantity
1    서울  마우스     25        40
4    대구  마우스     30        35
3    부산  키보드     80        12
7    광주  모니터    280        10
5    대구  모니터    300         8
2    부산  노트북   1150         5
0    서울  노트북   1200         3
6    광주  노트북   1300         2

[1-4] loc[0] -> 서울 노트북
[1-4] iloc[0] -> 서울 마우스
```

---

## 문제 2. 조건 추출과 정렬 (제한시간 10분)

- 문제 1의 `sales` 데이터를 계속 사용

#### 2-1. `quantity` 가 10 이상인 행을 **`query( )` 방식**과 **대괄호 방식** 두 가지로 각각 추출하시오.
#### 2-2. `store` 가 '서울' 또는 '부산' 이면서 `price` 가 100 이상인 행을 추출하시오.
#### 2-3. `revenue` 열(= `price` × `quantity`)을 추가한 뒤, `revenue` 기준 내림차순으로 정렬하고 인덱스를 0부터 다시 매기시오.

> 힌트 2-2: 대괄호 방식에서는 각 조건을 소괄호로 감싸야 함

In [ ]:
# 2-1. 두 가지 방식으로 동일 조건 추출
q1 = None              # TODO: query( ) 방식
q2 = None              # TODO: 대괄호 방식

# 2-2. 서울 또는 부산 이면서 price 100 이상
q3 = None              # TODO

# 2-3. revenue 열 추가 후 정렬 및 인덱스 재설정
sales2 = sales.copy()
sales2['revenue'] = None   # TODO
result = None              # TODO

print("[2-1] 두 방식의 결과가 같은가?", q1.equals(q2))
print(q1)
print()
print("[2-2]")
print(q3)
print()
print("[2-3]")
print(result)

**기대 출력**

```
[2-1] 두 방식의 결과가 같은가? True
  store item  price  quantity
1    서울  마우스     25        40
3    부산  키보드     80        12
4    대구  마우스     30        35
7    광주  모니터    280        10

[2-2]
  store item  price  quantity
0    서울  노트북   1200         3
2    부산  노트북   1150         5

[2-3]
  store item  price  quantity  revenue
0    부산  노트북   1150         5     5750
1    서울  노트북   1200         3     3600
2    광주  모니터    280        10     2800
3    광주  노트북   1300         2     2600
4    대구  모니터    300         8     2400
5    대구  마우스     30        35     1050
6    서울  마우스     25        40     1000
7    부산  키보드     80        12      960
```

---

## 문제 3. 치폴레 판매 기록 (제한시간 15분)

- 아래 셀을 먼저 실행할 것
- 치폴레(Chipotle) 매장의 주문 기록 데이터

In [1]:
import pandas as pd

chipo = pd.read_csv('https://raw.githubusercontent.com/justmarkham/DAT8/'
                    'master/data/chipotle.tsv', sep='\t')

print(chipo.shape)
print(chipo.head(3))
print()
print(chipo.dtypes)

(4622, 5)
   order_id  quantity  ... choice_description item_price
0         1         1  ...                NaN     $2.39 
1         1         1  ...       [Clementine]     $3.39 
2         1         1  ...            [Apple]     $3.39 

[3 rows x 5 columns]

order_id              int64
quantity              int64
item_name               str
choice_description      str
item_price              str
dtype: object


- `item_price` 의 자료형이 `object`(문자열)임에 주목할 것
  - `$8.75` 처럼 달러 기호가 붙어 있어 숫자로 인식되지 않음
  - 계산을 하려면 기호를 제거하고 자료형을 바꿔야 함

#### 3-1. `item_name` 이 `'Chicken Bowl'` 인 행만 추출하여 `bowl` 에 저장하시오. 원본은 그대로 둘 것.
#### 3-2. `bowl` 에 `price` 열을 추가하시오. `item_price` 에서 `$` 를 제거하고 실수형으로 변환한 값.
#### 3-3. `Chicken Bowl` 의 총 판매 금액, 총 판매 수량, 평균 판매 가격을 구하시오. 평균은 소수점 둘째 자리까지.

> 힌트 3-1: 조건으로 행을 추출한 뒤 사본을 만들 것
> 
> 힌트 3-2: 문자열 Series에 `.str.replace()` 함수로 기호 제거 가능. `.astype()` 함수로 자료형 변환 가능
> 
> 힌트 3-3: 총 판매 금액은 `price` 의 합, 총 판매 수량은 `quantity` 의 합

In [ ]:
# 3-1. Chicken Bowl 행만 추출 (사본)
bowl = None            # TODO

# 3-2. price 열 추가
bowl['price'] = None   # TODO

# 3-3. 총 금액, 총 수량, 평균 가격
total_price = None     # TODO
total_qty   = None     # TODO
avg_price   = None     # TODO

print("[3-1] 추출된 행 개수:", len(bowl))
print()
print("[3-2]")
print(bowl[['item_name', 'quantity', 'item_price', 'price']].head())
print()
print("[3-3] 총 판매 금액:", round(total_price, 2))
print("[3-3] 총 판매 수량:", total_qty)
print("[3-3] 평균 판매 가격:", avg_price)

**기대 출력**

```
[3-1] 추출된 행 개수: 726

[3-2]
       item_name  quantity item_price  price
4   Chicken Bowl         2    $16.98   16.98
5   Chicken Bowl         1    $10.98   10.98
13  Chicken Bowl         1    $11.25   11.25
19  Chicken Bowl         1     $8.75    8.75
26  Chicken Bowl         1     $8.49    8.49

[3-3] 총 판매 금액: 7342.73
[3-3] 총 판매 수량: 761
[3-3] 평균 판매 가격: 9.65
```

---

## 문제 4. EURO 2012 팀 기록 (제한시간 10분)

- 아래 셀을 먼저 실행할 것

In [1]:
import pandas as pd

euro = pd.read_csv('https://raw.githubusercontent.com/guipsamora/pandas_exercises/'
                   'master/02_Filtering_%26_Sorting/Euro12/Euro_2012_stats_TEAM.csv')

print(euro.shape)
print(euro[['Team', 'Goals', 'Shots on target', 'Passes', 'Passing Accuracy']])

(16, 35)
                   Team  Goals  Shots on target  Passes Passing Accuracy
0               Croatia      4               13    1076            76.9%
1        Czech Republic      4               13    1565            78.1%
2               Denmark      4               10    1298            83.3%
3               England      5               11    1488            80.6%
4                France      3               22    2066            87.2%
5               Germany     10               32    2774            87.4%
6                Greece      5                8    1187            76.7%
7                 Italy      6               34    3016            83.9%
8           Netherlands      2               12    1556            88.7%
9                Poland      2               15    1059            80.4%
10             Portugal      6               22    1891            77.2%
11  Republic of Ireland      1                7     851            71.2%
12               Russia      5            

#### 4-1. 결승에 진출한 두 팀(`Spain`, `Italy`)의 `Team`, `Goals`, `Shots on target` 을 추출하시오.
#### 4-2. 골을 가장 많이 넣은 상위 4개 팀을 구한 뒤, 그 팀들을 유효슈팅(`Shots on target`) 기준 내림차순으로 정렬하시오. `Team`, `Goals`, `Shots on target` 열만 출력할 것.
#### 4-3. 패스(`Passes`)가 가장 적은 5개 팀을 구한 뒤, 패스 정확도(`Passing Accuracy`) 기준 내림차순으로 정렬하시오.

> 힌트 4-1: 두 팀 이름으로 조건을 만들 것
> 
> 힌트 4-2: 정렬 후 `head()` 로 상위 팀을 얻고, 그 결과를 다시 정렬
> 
> 힌트 4-3: `Passing Accuracy` 는 `76.9%` 처럼 문자열이므로 기호 제거 및 자료형 변환 필요

In [ ]:
# 4-1. 결승 진출 두 팀
final2 = None          # TODO

# 4-2. 골 상위 4개 팀 -> 유효슈팅 내림차순
top4 = None            # TODO

# 4-3. 패스가 가장 적은 5개 팀 -> 패스 정확도 내림차순
low5 = None            # TODO: 여러 줄로 나누어 작성해도 됨

print("[4-1]")
print(final2)
print()
print("[4-2]")
print(top4)
print()
print("[4-3]")
print(low5)

**기대 출력**

```
[4-1]
     Team  Goals  Shots on target
7   Italy      6               34
13  Spain     12               42

[4-2]
        Team  Goals  Shots on target
13     Spain     12               42
7      Italy      6               34
5    Germany     10               32
10  Portugal      6               22

[4-3]
                   Team  Passes Passing Accuracy
14               Sweden    1192            80.9%
9                Poland    1059            80.4%
0               Croatia    1076            76.9%
6                Greece    1187            76.7%
11  Republic of Ireland     851            71.2%
```

---

## 문제 5. 출력을 보고 코드 복원하기 (제한시간 10분)

- 아래는 문제 1의 `sales` 데이터로 만든 출력임
- 각 출력을 만들어 내는 코드를 작성할 것
- 정답이 하나가 아닐 수 있으므로, 자신이 쓴 코드로 **같은 출력이 나오는지 반드시 실행하여 확인**할 것

> 힌트: 출력의 **인덱스 번호**를 먼저 볼 것.
> 인덱스가 0부터 순서대로면 원본 순서 그대로이고, 뒤섞여 있으면 정렬이나 추출을 거친 것임

#### 5-1. 다음 출력을 만드는 코드를 작성하시오.

```
  store item  price  quantity
2    부산  노트북   1150         5
0    서울  노트북   1200         3
6    광주  노트북   1300         2
```

#### 5-2. 다음 출력을 만드는 코드를 작성하시오.

```
store        대구
item        모니터
price       300
quantity      8
Name: 5, dtype: object
```

#### 5-3. 다음 출력을 만드는 코드를 작성하시오.

```
  item  price
0  노트북   1200
2  노트북   1150
5  모니터    300
6  노트북   1300
7  모니터    280
```